### ISSUE TITLE
Support profile engine builds for simulators

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#58988]
Today it is reported that `--ios --runtime-mode profile --simulator` can't be built, but an engine configured like this is needed for users who would like to run tests on AOT builds.

@mehmetf @chinmaygarde 

Separately, exposing this testing configuration to the Flutter tool would require building and packaging such an engine in the usual way on our infrastructure and removing certain gates from the Flutter tool. That work is not part of this issue, and should go in a separate issue if it's needed.

@jonahwilliams

**Comment by dnfield** [comment]
If the goal is to support AOT testing, why not do it via AOT mode in `flutter_tester` (currently not supported, but should be possible)?

**Comment by jonahwilliams** [comment]
`./flutter/tools/gn --simulator  --no-lto --simulator --runtime-mode=profile` works fine, possibly hitting the lack of LTO support for host builds?

**Comment by zanderso** [comment]
@dnfield I believe the request is specifically for driver based tests.

**Comment by dnfield** [comment]
@jonahwilliams  - you need the `--ios` arg in there as well, which will throw an error.

**Comment by jonahwilliams** [comment]
Solution: comment out 3 lines in flutter/gn
        # elif args.target_os == 'ios':
        #     if args.runtime_mode != 'debug':
        #         raise Exception('iOS simulator only supports the debug runtime mode')

**Comment by jonahwilliams** [comment]
Updates for the flutter tool (rough) to run ios profile mode:

```
diff --git a/examples/hello_world/ios/Flutter/.last_build_id b/examples/hello_world/ios/Flutter/.last_build_id
index 56cb05cd22..e9d2df3878 100644
--- a/examples/hello_world/ios/Flutter/.last_build_id
+++ b/examples/hello_world/ios/Flutter/.last_build_id
@@ -1 +1 @@
-8815a07f6ac6b55e6bd0d2350525ef75
\ No newline at end of file
+9ba788cbf5f041da5afd494b18e0e5e5
\ No newline at end of file
diff --git a/packages/flutter_tools/bin/xcode_backend.sh b/packages/flutter_tools/bin/xcode_backend.sh
index 98812435f7..e3309f840f 100755
--- a/packages/flutter_tools/bin/xcode_backend.sh
+++ b/packages/flutter_tools/bin/xcode_backend.sh
@@ -96,7 +96,7 @@ is set to release or run \"flutter build ios --release\", then re-run Archive fr
   local framework_path="${FLUTTER_ROOT}/bin/cache/artifacts/engine/${artifact_variant}"
   local flutter_engine_flag=""
   local local_engine_flag=""
-  local flutter_framework="${framework_path}/Flutter.framework"
+  local flutter_framework="${framework_path}/ "
   local flutter_podspec="${framework_path}/Flutter.podspec"
 
   if [[ -n "$FLUTTER_ENGINE" ]]; then
@@ -104,18 +104,18 @@ is set to release or run \"flutter build ios --release\", then re-run Archive fr
   fi
 
   if [[ -n "$LOCAL_ENGINE" ]]; then
-    if [[ $(echo "$LOCAL_ENGINE" | tr "[:upper:]" "[:lower:]") != *"$build_mode"* ]]; then
-      EchoError "========================================================================"
-      EchoError "ERROR: Requested build with Flutter local engine at '${LOCAL_ENGINE}'"
-      EchoError "This engine is not compatible with FLUTTER_BUILD_MODE: '${build_mode}'."
-      EchoError "You can fix this by updating the LOCAL_ENGINE environment variable, or"
-      EchoError "by running:"
-      EchoError "  flutter build ios --local-engine=ios_${build_mode}"
-      EchoError "or"
-      EchoError "  flutter build ios --local-engine=ios_${build_mode}_unopt"
-      EchoError "========================================================================"
-      exit -1
-    fi
+    # if [[ $(echo "$LOCAL_ENGINE" | tr "[:upper:]" "[:lower:]") != *"$build_mode"* ]]; then
+    #   EchoError "========================================================================"
+    #   EchoError "ERROR: Requested build with Flutter local engine at '${LOCAL_ENGINE}'"
+    #   EchoError "This engine is not compatible with FLUTTER_BUILD_MODE: '${build_mode}'."
+    #   EchoError "You can fix this by updating the LOCAL_ENGINE environment variable, or"
+    #   EchoError "by running:"
+    #   EchoError "  flutter build ios --local-engine=ios_${build_mode}"
+    #   EchoError "or"
+    #   EchoError "  flutter build ios --local-engine=ios_${build_mode}_unopt"
+    #   EchoError "========================================================================"
+    #   exit -1
+    # fi
     local_engine_flag="--local-engine=${LOCAL_ENGINE}"
     flutter_framework="${FLUTTER_ENGINE}/out/${LOCAL_ENGINE}/Flutter.framework"
     flutter_podspec="${FLUTTER_ENGINE}/out/${LOCAL_ENGINE}/Flutter.podspec"
diff --git a/packages/flutter_tools/lib/src/build_system/targets/ios.dart b/packages/flutter_tools/lib/src/build_system/targets/ios.dart
index cca692cae2..64ed124939 100644
--- a/packages/flutter_tools/lib/src/build_system/targets/ios.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/ios.dart
@@ -60,12 +60,12 @@ abstract class AotAssemblyBase extends Target {
     if (targetPlatform != TargetPlatform.ios) {
       throw Exception('aot_assembly is only supported for iOS applications.');
     }
-    if (iosArchs.contains(DarwinArch.x86_64)) {
-      throw Exception(
-        'release/profile builds are only supported for physical devices. '
-        'attempted to build for $iosArchs.'
-      );
-    }
+    // if (iosArchs.contains(DarwinArch.x86_64)) {
+    //   throw Exception(
+    //     'release/profile builds are only supported for physical devices. '
+    //     'attempted to build for $iosArchs.'
+    //   );
+    // }
 
     // If we're building multiple iOS archs the binaries need to be lipo'd
     // together.
diff --git a/packages/flutter_tools/lib/src/commands/assemble.dart b/packages/flutter_tools/lib/src/commands/assemble.dart
index 62c46b2d15..db1ec933b2 100644
--- a/packages/flutter_tools/lib/src/commands/assemble.dart
+++ b/packages/flutter_tools/lib/src/commands/assemble.dart
@@ -216,6 +216,7 @@ class AssembleCommand extends FlutterCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
+    print(argResults.arguments);
     final List<Target> targets = createTargets();
     final Target target = targets.length == 1 ? targets.single : _CompositeTarget(targets);
     final BuildResult result = await globals.buildSystem.build(
diff --git a/packages/flutter_tools/lib/src/ios/simulators.dart b/packages/flutter_tools/lib/src/ios/simulators.dart
index dc9bdf8d28..9432800b6c 100644
--- a/packages/flutter_tools/lib/src/ios/simulators.dart
+++ b/packages/flutter_tools/lib/src/ios/simulators.dart
@@ -311,7 +311,7 @@ class IOSSimulator extends Device {
   Future<bool> get supportsHardwareRendering async => false;
 
   @override
-  bool supportsRuntimeMode(BuildMode buildMode) => buildMode == BuildMode.debug;
+  bool supportsRuntimeMode(BuildMode buildMode) => true;
 
   Map<ApplicationPackage, _IOSSimulatorLogReader> _logReaders;
   _IOSSimulatorDevicePortForwarder _portForwarder;
```


Updates to engine repo required:

```
diff --git a/tools/gn b/tools/gn
index 8e071997b..f5b284330 100755
--- a/tools/gn
+++ b/tools/gn
@@ -65,9 +65,9 @@ def to_gn_args(args):
     if args.simulator:
         if args.target_os == 'android':
             raise Exception('--simulator is not supported on Android')
-        elif args.target_os == 'ios':
-            if args.runtime_mode != 'debug':
-                raise Exception('iOS simulator only supports the debug runtime mode')
+        # elif args.target_os == 'ios':
+        #     if args.runtime_mode != 'debug':
+        #         raise Exception('iOS simulator only supports the debug runtime mode')
 
     if args.target_os != 'android' and args.enable_vulkan:
       raise Exception('--enable-vulkan is only supported on Android')
```

For a local engine build, this requires us to symlink gen_snapshot to gen_snapshot_x86_64

**Comment by chinmaygarde** [comment]
This seems fairly low risk to me as the unit-test harnesses already exercise x64 AOT code paths on the host. All restrictions to building such engines are artificial.

> possibly hitting the lack of LTO support for host builds?

This should be supported but is not used because of build capacity constraints. Even if you don't end up LTO-ing this new variant, can you file a bug to track chasing down a fix for the broken host LTO builds please?

> For a local engine build, this requires us to symlink gen_snapshot to gen_snapshot_x86_64

This is a known issue and present on all variants https://github.com/flutter/flutter/issues/38933.

**Comment by jonahwilliams** [comment]
I recall @dnfield mentioned the LTO issue, I haven't investigated. Pending engine capacity we could upload profile/release builds, but until then it seems like g3 is unblocked
